File: examples/python/linkcampaignimpl/memory_repo.py
from dataclasses import dataclass

from campaign.campaign import Campaign, CampaignSpec
from campaign.campaign_id import CampaignID
from campaign.short_link import ShortLinkSpec


@dataclass(frozen=True)
class _LinkRecord:

    slug: str
    target_url: str
    active: bool


@dataclass(frozen=True)
class _CampaignRecord:
    id: str
    name: str
    links: tuple[_LinkRecord, ...]

    def to_spec(self) -> CampaignSpec:
        return CampaignSpec(
            id=self.id,
            name=self.name,
            links=tuple(
                ShortLinkSpec(slug=l.slug, target_url=l.target_url, active=l.active)
                for l in self.links
            ),
        )


class InMemoryCampaignRepository:

    def __init__(self) -> None:
        self._campaigns: dict[str, _CampaignRecord] = {}

    def save(self, c: Campaign) -> None:
        self._campaigns[str(c.id)] = _decompose(c)

    def load(self, id: CampaignID) -> Campaign:
        rec = self._campaigns.get(str(id))
        if rec is None:
            raise LookupError(f"campaign {id} not found")
        return Campaign(rec.to_spec())


def _decompose(c: Campaign) -> _CampaignRecord:
    return _CampaignRecord(
        id=str(c.id),
        name=str(c.name),
        links=tuple(
            _LinkRecord(slug=str(l.slug), target_url=str(l.target_url), active=l.active)
            for l in c.links
        ),
    )

File: examples/running/linkcampaignimpl/memory_repo.go
package linkcampaignimpl

import (
	"context"
	"fmt"
	"sync"

	"github.com/verocorp/tesser-build/examples/running/campaign"
)

type campaignRecord struct {
	id    string
	name  string
	links []linkRecord
}

type linkRecord struct {
	slug      string
	targetURL string
	active    bool
}

type InMemoryCampaignRepository struct {
	mu        sync.Mutex
	campaigns map[string]campaignRecord
}

func NewInMemoryCampaignRepository() *InMemoryCampaignRepository {
	return &InMemoryCampaignRepository{campaigns: make(map[string]campaignRecord)}
}

func (r *InMemoryCampaignRepository) Save(_ context.Context, c campaign.Campaign) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.campaigns[c.ID().String()] = decompose(c)
	return nil
}

func (r *InMemoryCampaignRepository) Load(_ context.Context, id campaign.CampaignID) (campaign.Campaign, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	rec, ok := r.campaigns[id.String()]
	if !ok {
		return campaign.Campaign{}, fmt.Errorf("campaign %s not found", id)
	}
	return campaign.NewCampaign(rec.toSpec())
}

func decompose(c campaign.Campaign) campaignRecord {
	links := c.Links()
	linkRecs := make([]linkRecord, len(links))
	for i, l := range links {
		linkRecs[i] = linkRecord{slug: l.Slug().String(), targetURL: l.TargetURL().String(), active: l.Active()}
	}
	return campaignRecord{id: c.ID().String(), name: c.Name().String(), links: linkRecs}
}

func (rec campaignRecord) toSpec() campaign.CampaignSpec {
	links := make([]campaign.ShortLinkSpec, len(rec.links))
	for i, l := range rec.links {
		links[i] = campaign.ShortLinkSpec{Slug: l.slug, TargetURL: l.targetURL, Active: l.active}
	}
	return campaign.CampaignSpec{ID: rec.id, Name: rec.name, Links: links}
}
